ARROW-10174: [Java] Fix reading/writing dict structs

HedgehogCode opened the pull request:
When translating between the memory FieldType and message FieldType for
dictionary encoded vectors the children of the dictionary field were not
handled correctly.
* When going from memory format to message format the Field must have the
  children of the dictionary field.
* When going from message format to memory format the Field must have no
  children but the dictionary must have the mapped children

emkornfield commented:
@liyafan82 do you have time to review?

liyafan82 commented:
> @liyafan82 do you have time to review?

@emkornfield Sure. I will take a look in one or two days.

liyafan82 commented on java/vector/src/main/java/org/apache/arrow/vector/util/DictionaryUtility.java:
```diff
@@ -115,25 +118,28 @@ public static Field toMemoryFormat(Field field, BufferAllocator allocator, Map<L
     } else {
       // re-type the field for in-memory format
       type = encoding.getIndexType();
+      fieldChildren = null;
```
This seems not necessary?

HedgehogCode replied:
Correct. I moved the `= null` to the declaration of `fieldChildren`. (Usually I would make `fieldChildren` final and then this would be necessary. However, this seems not to be in line with the general code style in the repository).

liyafan82 commented on java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java:
```diff
@@ -305,6 +321,64 @@ public void testWriteReadWithDictionaries() throws IOException {
+    values.put("a", Arrays.asList(1, 3, 2, 3, 2, 1, 1));
+    values.put("b", Arrays.asList(4, 6, 5, 6, 5, 4, 4));
+    setVector(vector, values);
+    FieldVector encodedVector = (FieldVector) DictionaryEncoder.encode(vector, dictionary4);
```
Maybe we need to verify the encoded vector?

HedgehogCode replied:
No. Encoding of struct vectors is tested in `TestDictionaryVector#testEncodeStruct`. The encoded vector should be fine.

liyafan82 commented on java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java:
```diff
@@ -305,6 +321,64 @@ public void testWriteReadWithDictionaries() throws IOException {
+        assertEquals(1, reader.getDictionaryVectors().size());
+        
+        final FieldVector readEncoded = readRoot.getVector(0);
+        assertTrue(readEncoded instanceof IntVector);
```
should we validate `readEncoded`?

HedgehogCode replied:
Yes. It is interesting to check the `Field`. However, using the `RangeEqualsVisitor` we can check both without any more code. So I did it.

liyafan82 commented on java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java:
```diff
@@ -305,6 +321,64 @@ public void testWriteReadWithDictionaries() throws IOException {
+        final Map<Long, Dictionary> readDictionaryMap = reader.getDictionaryVectors();
+        final Dictionary readDictionary =
+            readDictionaryMap.get(readEncoded.getField().getDictionary().getId());
+        assertNotNull(readDictionary);
```
should we validate the read dictionary vector?

HedgehogCode replied:
Yes. I added a check.

liyafan82 commented on java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java:
```diff
@@ -305,6 +321,64 @@ public void testWriteReadWithDictionaries() throws IOException {
+        assertEquals(vector.getValueCount(), readVector.getValueCount());
+        for (int i = 0; i < vector.getValueCount(); i++) {
+          assertEquals(vector.getObject(i), readVector.getObject(i));
+        }
```
Maybe we can compare the two vectors through a `RangeEqualsVisitor` object?

HedgehogCode replied:
Good advice! The `RangeEqualsVisitor` does some checks on the `Field` that were missing before.

liyafan82 commented on java/vector/src/test/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java:
```diff
@@ -673,4 +677,28 @@ public static void setVector(FixedSizeListVector vector, List<Integer>... values
+      child.allocateNew();
+      final List<Integer> v = entry.getValue();
+      for (int i = 0; i < v.size(); i++) {
+        child.set(i, v.get(i));
```
We need to consider the null values in v?

HedgehogCode replied:
Yes, why not. I added an `if` switch for them.

liyafan82 commented on java/vector/src/test/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java:
```diff
@@ -673,4 +677,28 @@ public static void setVector(FixedSizeListVector vector, List<Integer>... values
+      for (int i = 0; i < v.size(); i++) {
+        child.set(i, v.get(i));
+        vector.setIndexDefined(i);
+      }
```
Maybe we also need to set the value count for the child vector

HedgehogCode replied:
No. The struct vector sets the value count of its children in `#setValueCount`.

HedgehogCode commented:
I implemented the changes and force pushed them.

Thank you for the advice with the `RangeEqualsVisitor`. Using it I discovered another bug in `DictionaryUtility`. The field for the Dictionary was always created with `nullable=false` but this is incorrect because dictionaries can be nullable. Therefore, I replaced this with the nullability read from the schema.

liyafan82 commented on java/vector/src/test/java/org/apache/arrow/vector/testing/ValueVectorDataPopulator.java:
```diff
@@ -673,4 +677,32 @@ public static void setVector(FixedSizeListVector vector, List<Integer>... values
+        } else {
+          child.setNull(i);
+        }
+        vector.setIndexDefined(i);
```
we should call `setIndexDefined` only if `v.get(i) != null`?

HedgehogCode replied:
I am not sure.
The current implementation does not allow setting any struct element to `null`. If we only call `#setIndexDefined` if the child element is not `null` the implementation would not allow (non null) structs with all child values beeing `null`. I am okay with both but I think we should keep the API simple because this method is just there to make tests shorter.

liyafan82 replied:
For a StructVector, we can set an element to null. For its super class (NonNullableStructVector), we cannot set an element to null. 

Here, I generally prefer setting an element to null if all sub-elements are null, because ValueVectorDataPopulator is a generally-purpose class, and we may use it to test scenarios with null elements.

HedgehogCode replied:
I see. I changed it.

liyafan82 commented on java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java:
```diff
@@ -305,6 +325,77 @@ public void testWriteReadWithDictionaries() throws IOException {
+                .rangeEquals(new Range(0, 0, vector.getValueCount())));
+
+        vector.close();
+        readVector.close();
```
it would be beneficial to close them through a try-with-resource clause to avoid resource leak?

HedgehogCode replied:
Yes, I changed it. I did not do it before because I did not want the deep nesting.

liyafan82 replied:
I see. Deep nesting is not good-looking.

liyafan82 reviewed (approved):
LGTM, will merge soon if there are no more comments.

liyafan82 commented:
Merging. Thanks for your effort. @HedgehogCode
